### PR TITLE
Ray casting in scene editor

### PR DESCRIPTION
**Pull request by Erdios**
Is built based on:
1. Main brunch.

What changes:
1. Add a new constructor in Matrix class.
2. Add basic ray casting functionality in scene editor (cannot highlight the object yet)
3. Improve UI performance in Hierarchy Window.

### CODE REVIEW

File: Source/SceneEditor/EditorCamera.cpp [hunk 1]
```diff
@@ -141,50 +145,95 @@ namespace PlatinumEngine
...
 		// update euler angle
 		_eulerAngle += eulerAngle;
 
-		PlatinumEngine::Maths::Mat4 rotationMat4, translationMat4, translation2Mat4;
-
-		rotationMat4.SetRotationMatrix(-_eulerAngle);
+		PlatinumEngine::Maths::Mat4 rotationMat4, translationMat4, rotationMat4Inverse, translationMat4Inverse;
 
+		// get rotation matrix
+		rotationMat4Inverse.SetRotationMatrix(-_eulerAngle);
```
Comment (kablouser) [inline Source/SceneEditor/EditorCamera.cpp:156]: mathematics muhahaha

File: Source/SceneEditor/SceneEditor.cpp [hunk 2]
```diff
@@ -264,6 +251,311 @@ namespace PlatinumEngine{
...
+	{
+		//-----------------//
+		// Calculate ray   //
+		//-----------------//
+		PlatinumEngine::Maths::Vec3 ray;
+
+		// check projection type
+		if(_camera.isOrthogonal)
+		{
+			ray = PlatinumEngine::Maths::Vec3(0.f, 0.f, 1.f);
```
Comment (kablouser) [inline Source/SceneEditor/SceneEditor.cpp:281]: Temporary I hope?

File: Source/SceneEditor/SceneEditor.cpp [hunk 3]
```diff
@@ -264,6 +251,311 @@ namespace PlatinumEngine{
...
+		}
+
+
+		//----------------------------------------//
+		// Loop through Every root game object    //
+		//----------------------------------------//
+
+		// variable that store the current selected game object
+		GameObject* currentSelectedGameobject = nullptr;
+		float closetZValue = (float)-_far;
```
Comment (kablouser) [inline Source/SceneEditor/SceneEditor.cpp:315]: This is correct in right handed system.
But it's +_far in left handed system.

File: Source/SceneEditor/SceneEditor.cpp [hunk 4]
```diff
@@ -264,6 +251,311 @@ namespace PlatinumEngine{
...
+									PlatinumEngine::Maths::Vec4(crossPoint.x, crossPoint.y, crossPoint.z, 1.0f);
+
+							//crossPointViewCoordinate = crossPointViewCoordinate/crossPointViewCoordinate.w;
+
+							// This part check
+							// check if the current cross point has the closest z value.
+							// check if the current cross point is within clipping space.
+							if (closestZValueForCrossPoint < crossPointViewCoordinate.z &&
+								crossPointViewCoordinate.z >= (float)-_far &&
+								crossPointViewCoordinate.z <= (float)-_near)
```
Comment (kablouser) [inline Source/SceneEditor/SceneEditor.cpp:530]: Correct in right handed system.
But left handed is *different*.

File: Source/SceneEditor/SceneEditor.cpp [hunk 5]
```diff
@@ -264,6 +251,311 @@ namespace PlatinumEngine{
...
+						vertex0 = mesh->vertices[mesh->indices[count + 0]].position;
+						vertex1 = mesh->vertices[mesh->indices[count + 1]].position;
+						vertex2 = mesh->vertices[mesh->indices[count + 2]].position;
+					}
+
+					//----------------------------------//
+					// Do ray and triangle intersection //
+					//----------------------------------//
+
+					// build the PCS using the triangle primitive
```
Comment (kablouser) [inline Source/SceneEditor/SceneEditor.cpp:413]: What is PCS? Maybe add link for algorithm? This looks unfamiliar.

File: Source/SceneEditor/SceneEditor.cpp [hunk 6]
```diff
@@ -264,6 +251,311 @@ namespace PlatinumEngine{
...
+
+					// origin of PCS
+					Maths::Vec3 originForPCS = vertex0;
+
+					// not normalized axis of PCS
+					Maths::Vec3 axisUForPCS = vertex1 - originForPCS;
+					Maths::Vec3 axisNForPCS = Cross(axisUForPCS, (vertex2 - originForPCS));
+					Maths::Vec3 axisWForPCS = Cross(axisUForPCS, axisNForPCS);
```
Comment (kablouser) [inline Source/SceneEditor/SceneEditor.cpp:421]: Axis U N W?

File: Source/SceneEditor/SceneEditor.cpp [hunk 7]
```diff
@@ -264,6 +251,311 @@ namespace PlatinumEngine{
...
+					axisWForPCS = axisWForPCS / lengthW;
+
+
+
+					// find cross point
+
+					// find the intersection point for ray and the panel that contains the triangle
+					Maths::Vec3 vectorBetweenOriginForPCSandCameraPosition = originForPCS - inCameraPosition;
+
+					float rateForLineApproachingPanel = Maths::Dot(inRay, axisNForPCS);
```
Comment (kablouser) [inline Source/SceneEditor/SceneEditor.cpp:447]: Rate for line? Just call it dot.

File: Source/SceneEditor/SceneEditor.cpp [hunk 8]
```diff
@@ -264,6 +251,311 @@ namespace PlatinumEngine{
...
+					//------------------------------------//
+					// Do barycentric interpolation check //
+					//------------------------------------//
+
+					// Convert the cross point and the vertices of the triangle into PCS coordinate
+					Maths::Mat3 PCSMatrix3x3(std::array<float, 9>({
+							axisUForPCS.x, axisWForPCS.x, axisNForPCS.x,
+							axisUForPCS.y, axisWForPCS.y, axisNForPCS.y,
+							axisUForPCS.z, axisWForPCS.z, axisNForPCS.z
+					}).data());
```
Comment (kablouser) [inline Source/SceneEditor/SceneEditor.cpp:473]: Looks like Mat3 needs more constructors.

File: Source/SceneEditor/SceneEditor.cpp [hunk 9]
```diff
@@ -264,6 +251,311 @@ namespace PlatinumEngine{
...
+
+					// not normalized axis of PCS
+					Maths::Vec3 axisUForPCS = vertex1 - originForPCS;
+					Maths::Vec3 axisNForPCS = Cross(axisUForPCS, (vertex2 - originForPCS));
+					Maths::Vec3 axisWForPCS = Cross(axisUForPCS, axisNForPCS);
+
+					// length of axis
+					float lengthU = sqrtf(LengthSquared(axisUForPCS));
+					float lengthN = sqrtf(LengthSquared(axisNForPCS));
+					float lengthW = sqrtf(LengthSquared(axisWForPCS));
```
Comment (kablouser) [inline Source/SceneEditor/SceneEditor.cpp:426]: Length()?

File: Source/SceneEditor/SceneEditor.cpp [hunk 10]
```diff
@@ -264,6 +251,311 @@ namespace PlatinumEngine{
...
+
+					// 2. triangle vertices
+					Maths::Vec3 vertex0PCS = PCSMatrix3x3 * (vertex0 - originForPCS);
+					Maths::Vec3 vertex1PCS = PCSMatrix3x3 * (vertex1 - originForPCS);
+					Maths::Vec3 vertex2PCS = PCSMatrix3x3 * (vertex2 - originForPCS);
+
+					// Do barycentric interpolation to check
+					Maths::Vec3 xValueOfVerticesPCS((vertex0PCS - vertex1PCS).x, (vertex0PCS - vertex2PCS).x,
+							(crossPointPCS - vertex0PCS).x);
+					float lengthOfXValueOfVerticesPCS = sqrtf(LengthSquared(xValueOfVerticesPCS));
```
Comment (kablouser) [inline Source/SceneEditor/SceneEditor.cpp:485]: Length()?

File: Source/SceneEditor/SceneEditor.cpp [hunk 11]
```diff
@@ -264,6 +251,311 @@ namespace PlatinumEngine{
...
+
+					Maths::Vec3 yValueOfVerticesPCS((vertex0PCS - vertex1PCS).y, (vertex0PCS - vertex2PCS).y,
+							(crossPointPCS - vertex0PCS).y);
+					float lengthOfYValueOfVerticesPCS = sqrtf(LengthSquared(yValueOfVerticesPCS));
```
Comment (kablouser) [inline Source/SceneEditor/SceneEditor.cpp:489]: Length()?

**Review by kablouser — COMMENTED** [review]
Raycasting algorithm needs comment to a link or something explaining how it works. Google doesn't know PCS.

File: Source/SceneEditor/SceneEditor.cpp (hunk 6 above)
Comment (Erdios) [reply Source/SceneEditor/SceneEditor.cpp:421]: Mentioned in intersection PowerPoint for fundamental modeling last semester.

File: Source/SceneEditor/SceneEditor.cpp (hunk 5 above)
Comment (Erdios) [reply Source/SceneEditor/SceneEditor.cpp:413]: Mentioned in intersection PowerPoint for fundamental modeling last semester.

File: Source/SceneEditor/EditorCamera.cpp (hunk 1 above)
Comment (Erdios) [reply Source/SceneEditor/EditorCamera.cpp:156]: muhahaha, this will be changed in gizmo branch.
do i need to do it here?

**Review by ShawnWicky — APPROVED** [review]
💙 💜 💚 ❤️ 💛

**Pull request merged**